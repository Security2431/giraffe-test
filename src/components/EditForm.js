/* eslint no-underscore-dangle: 0 */
/* eslint jsx-a11y/label-has-for: 0 */
/* eslint jsx-a11y/label-has-associated-control: 0 */
import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import cuid from 'cuid';

class EditForm extends React.Component {
  static propTypes = {
    activeUser: PropTypes.string.isRequired,
    ad: PropTypes.shape({
      _id: PropTypes.string.isRequired,
      author_name: PropTypes.string.isRequired,
      created_at_datetime: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
    }),
    onSubmit: PropTypes.func.isRequired,
  };

  static defaultProps = {
    ad: null,
  };

  state = {
    title: {
      value: '',
    },
    description: {
      value: '',
    },
  };

  componentWillReceiveProps(nextProps) {
    if (!nextProps.ad) {
      this.setState({
        title: {
          value: '',
        },
        description: {
          value: '',
        },
      });
    }

    // eslint-disable-next-line
    if (nextProps.ad && nextProps.ad !== this.props.ad) {
      this.setState({
        title: {
          value: nextProps.ad.title,
        },
        description: {
          value: nextProps.ad.description,
        },
      });
    }
  }

  handleInputChange = (event) => {
    event.persist();
    const { name, value } = event.target;

    this.setState(prevState => ({
      [name]: {
        ...prevState[name],
        value,
      },
    }));
  };

  handleSubmit = (event) => {
    event.preventDefault();

    const { title, description } = this.state;
    const { activeUser, onSubmit, ad } = this.props;

    // eslint-disable-next-line
    const id = !!ad ? ad._id : cuid();

    const params = {
      ...ad,
      _id: id,
      author_name: activeUser,
      created_at_datetime: moment().format(),
      title: title.value,
      description: description.value,
    };

    onSubmit(params);
  };

  render() {
    const { title, description } = this.state;
    const { ad } = this.props;

    return (
      <div className="col-md-6 col-lg-5">
        <form className="form" onSubmit={this.handleSubmit}>
          <div className="form-group row">
            <label className="col-sm-2" htmlFor="title">
              Title
            </label>
            <div className="col-sm-10">
              <input
                className="form-control"
                id="title"
                type="text"
                name="title"
                autoComplete="title"
                value={title.value}
                onChange={this.handleInputChange}
                required
              />
            </div>
          </div>
          <div className="form-group row">
            <label className="col-sm-2" htmlFor="description">
              Description
            </label>
            <div className="col-sm-10">
              <textarea
                className="form-control"
                id="description"
                name="description"
                rows="8"
                value={description.value}
                autoComplete="description"
                onChange={this.handleInputChange}
                required
              />
            </div>
          </div>
          <div className="form-group row">
            <div className="col-sm-10 offset-sm-2">
              <button className="btn btn-primary btn-block" type="submit">
                {ad ? 'Edit' : 'Create'}
              </button>
            </div>
          </div>
        </form>
      </div>
    );
  }
}

export default EditForm;
