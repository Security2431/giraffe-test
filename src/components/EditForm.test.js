/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import EditForm from './EditForm';

jest.mock('cuid');

// Return a fixed timestamp when moment().fromNow() is called
jest.mock('moment', () => () => ({ format: () => '' }));

const mockProps = {
  activeUser: 'Admin',
  ad: {
    _id: 'cjv7wpcgj00022y5rtui74uls',
    author_name: 'Admin',
    created_at_datetime: '2019-05-03T12:59:23+03:00',
    title: 'Title',
    description: 'Hello, World!',
  },
  onSubmit: jest.fn(),
};

describe('<EditForm />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<EditForm {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<EditForm {...mockProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
