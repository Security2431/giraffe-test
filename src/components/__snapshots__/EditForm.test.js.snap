// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditForm /> renders correctly 1`] = `
<div
  className="col-md-6 col-lg-5"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form-group row"
    >
      <label
        className="col-sm-2"
        htmlFor="title"
      >
        Title
      </label>
      <div
        className="col-sm-10"
      >
        <input
          autoComplete="title"
          className="form-control"
          id="title"
          name="title"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="form-group row"
    >
      <label
        className="col-sm-2"
        htmlFor="description"
      >
        Description
      </label>
      <div
        className="col-sm-10"
      >
        <textarea
          autoComplete="description"
          className="form-control"
          id="description"
          name="description"
          onChange={[Function]}
          required={true}
          rows="8"
          value=""
        />
      </div>
    </div>
    <div
      className="form-group row"
    >
      <div
        className="col-sm-10 offset-sm-2"
      >
        <button
          className="btn btn-primary btn-block"
          type="submit"
        >
          Edit
        </button>
      </div>
    </div>
  </form>
</div>
`;
